compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -11465,43 +11465,45 @@ OMR::Z::TreeEvaluator::loadaddrEvaluator(TR::Node * node, TR::CodeGenerator * cg
    if (symRef->isUnresolved())
       {
 #ifdef J9_PROJECT_SPECIFIC
-      if (cg->getS390ProcessorInfo()->supportsArch(TR_S390ProcessorInfo::TR_z10))
-         {
-         TR::UnresolvedDataSnippet * uds;
-         TR::Instruction * cursor;
+      // create UnresolvedDataSnippet
+      TR::UnresolvedDataSnippet * uds = new (cg->trHeapMemory()) TR::UnresolvedDataSnippet(cg, node, symRef, false, false);
+      cg->addSnippet(uds);
 
-         // create UnresolvedDataSnippet
-         uds = new (cg->trHeapMemory()) TR::UnresolvedDataSnippet(cg, node, symRef, false, false);
-         cg->addSnippet(uds);
+      // generate branch to the unresolved data snippet
+      TR::Instruction * cursor = generateRegUnresolvedSym(cg, TR::InstOpCode::getLoadOpCode(), node, targetRegister, symRef, uds);
+      uds->setBranchInstruction(cursor);
 
-         // generate branch to the unresolved data snippet
-         cursor = generateRegUnresolvedSym(cg, TR::InstOpCode::getLoadOpCode(), node, targetRegister, symRef, uds);
+      // create a patchable data in litpool
+      TR::S390WritableDataSnippet * litpool = cg->CreateWritableConstant(node);
+      litpool->setUnresolvedDataSnippet(uds);
 
-         uds->setBranchInstruction(cursor);
-
-         // create a patchable data in litpool
-         TR::S390WritableDataSnippet * litpool = cg->CreateWritableConstant(node);
-         litpool->setUnresolvedDataSnippet(uds);
-
-         TR::S390RILInstruction * LRLinst;
-         LRLinst = (TR::S390RILInstruction *) generateRILInstruction(cg, TR::InstOpCode::getLoadRelativeLongOpCode(), node, targetRegister, reinterpret_cast<void*>(0xBABE), 0);
+      TR::S390RILInstruction * LRLinst;
+      if (cg->getS390ProcessorInfo()->supportsArch(TR_S390ProcessorInfo::TR_z10))
+         {
+         LRLinst = static_cast<TR::S390RILInstruction *>(generateRILInstruction(cg, TR::InstOpCode::getLoadRelativeLongOpCode(), node, targetRegister, reinterpret_cast<void*>(0xBABE), 0));
```

Comment:
There is a reason `generateRILInstruction` does not return a `TR::S390RILInstruction*` type. In general one should not assume this because arbitrary instructions may be inserted before the instruction you wanted to generate, and sometimes even after. If you want to do this case I'd like to see a fatal assert validating that `LRinst` is of the correct instruction kind by checking via `getKind`. Similarly for the `else` path.

Reply:
Added a fatal assert to check for instruction format.